### ISSUE TITLE
Rewrite adaptive_mutation as a real evolution example with a fitness function

### ISSUE THREAD

**Issue by stservice** [stSoftwareAU/NEAT-AI-Examples#254]
## Background

Issue #212 asked to audit `adaptive_mutation` per the framing of #203 (minimal `(input, output)` seed, `evolveDir` over a binary training set, per-generation `best_fitness`/`mean_fitness` CSV, `targetError` + `timeoutMinutes: 5` stop conditions, demonstrate the final creature solves the problem).

That framing is incompatible with what `adaptive_mutation` currently is. The demo:

- Has **no fitness function, no training data, no NEAT-AI training pass**. The header comment and README both state explicitly: *"No actual NEAT-AI training pass is invoked. The 'evolution' here is a deterministic mutation walk that updates each creature's neuron and synapse counts after each operator."*
- Compares a **small seed (~5 hidden)** against a **large seed (~256 hidden, ~10k synapses)** — the two seeds *are* the demo.
- Is explicitly listed as exempt in `AGENTS.md` under *"Exempt examples (hand-crafted state IS the demo)"* — *"`adaptive_mutation` — visualises the mutation operator distribution on a pre-built creature."*

**This design is wrong.** Every example — batch or event-driven — must have a fitness function. Without one there is no way to measure whether the population is improving. The purpose of these examples is to demonstrate how the library solves common AI/evolution problems; the AI should find the solution itself, not be given the answer. If the AI cannot find a good solution to a well-known problem, that is a genuine limitation worth raising as a separate issue.

## Decision

**Rewrite `adaptive_mutation` into a real evolution example** that satisfies #212's acceptance criteria:

- Define a concrete task (e.g. categorising a binary pattern set — problems like handwritten digit classification should be straightforward for this implementation).
- Generate binary training data and call `Creature.evolveDir()` to run a population through generations of evolution.
- Implement a proper fitness function so improvement is measurable.
- Record per-generation `best_fitness`/`mean_fitness` to a CSV.
- Use `targetError` + `timeoutMinutes: 5` stop conditions.
- Demonstrate that the final creature solves the problem.
- Remove the `AGENTS.md` exemption for `adaptive_mutation` once the rewrite is done.

If gathering the required per-generation statistics requires new options in NEAT-AI, raise those as separate issues in that repo.

> **Note on event-driven examples:** Event-driven support is currently limited and requires too much boilerplate in examples. Issues have been raised in NEAT-AI to address this. Until those are resolved, prefer batch (`evolveDir`) examples.

## Related

- Parent: #203 (closed) — directory list included `adaptive_mutation`.
- Sub-issue: #212 — closed in favour of this decision-issue.
- `AGENTS.md` — exempt-examples list and no-warm-start policy.

**Comment by nleck** [comment]
We're trying to demonstrate how our library solves problems, we're giving common problems that AI learning/evolution approaches can solve.

I absolutely don't want you to give the examples the answer, the AI should be able to find a good solution. 

If the AI can't find a solution to a common problem then that's a limitation with our solution and GH issues should be raised to get that fixed, that my learnings and that's a good thing.

Problems like "categorising handwritten digits" should be extremely easy for our implementation, generate binary training data, then call Creature.evolveDir() this will create a population of creatures and go through generations of evolution to find a good solution 

We may want to add a few options to NEAT-AI to gather and return the required statistics. 

We have discovered that event driven isn't as well supported ( not really at all) you needed to do too much coding in the examples. You have raised issues in NEAT-AI to address that

**Comment by nleck** [comment]
"No fitness function" ??? Every example event or batch needs a fitness function!!!

How would you know if we're getting better or worse if not? 

I don't understand what you're trying to say

**Comment by stservice** [comment]
Claimed by `stservice-1778367342031`

**Comment by stservice** [comment]
⏳ **Refinement in progress** — Processing 2 comment(s) from nleck. The issue will be updated shortly.

**Comment by stservice** [comment]
## Issue Refinement

The user feedback makes clear that (1) every example must have a fitness function — the current 'no fitness function' framing is fundamentally wrong, not a valid design choice — and (2) examples must demonstrate the library actually solving a problem via evolution, not just visualising internal state. This resolves the 'decision needed' framing: option 1 (keep as policy visualisation) is off the table. The title is updated to reflect the chosen direction, and the body is rewritten to treat the rewrite as decided, explain why the current design violates core example principles, and include concrete acceptance criteria and a note about event-driven limitations.

**Changes:** title updated, body updated

---
🤖 Processed by: stservice

**Comment by stservice** [comment]
Automatic recovery: this issue was unassigned from `stservice` because the worker has no active heartbeat for this issue (assigned over 30 minutes ago with no heartbeat recorded). This likely means the worker crashed between claiming the issue and starting work. The issue is now available to be picked up again.

---

Assigned-without-heartbeat recovery (Issue #632)

**Comment by stservice** [comment]
Claimed by `stservice-1778371412628`

**Comment by stservice** [comment]
## Planning complete — 3 sub-issues created

The rewrite has been broken into three sub-issues that land in dependency order. Each is sized for a single PR (≤ ~500 LOC, ≤ 5 files modified).

| # | Sub-issue | Scope |
| - | --------- | ----- |
| #262 | adaptive_mutation: add binary classification task primitives (data + fitness + tests) | Foundational: new `classification_task.ts` + tests; no changes to the existing example yet |
| #263 | adaptive_mutation: rewire evolution loop to use the classification task | Replaces the synthetic-regression imports in `adaptive_mutation.ts` with the new primitives, updates tests + CSV schema |
| #264 | adaptive_mutation: refresh README, charts, and AGENTS.md exemption for the classification rewrite | Updates `svg.ts`, `README.md`, regenerates artefacts, moves the example out of the AGENTS.md exempt list into the in-scope list |

### Suggested implementation order

1. #262 — add classification primitives (no impact on existing example)
2. #263 — wire them into the evolution loop (existing SVGs + README continue to render with stale labels)
3. #264 — refresh the visuals, README narrative, and AGENTS.md classification

### Dependency graph

```mermaid
flowchart LR
    A["#262 Classification primitives<br/>(data + fitness + tests)"]
    B["#263 Rewire evolution loop"]
    C["#264 README / SVGs / AGENTS.md"]

    A --> B --> C
```

### Decisions made during planning

- **Recommended task: 4-bit even parity.** Well-known generalisation of XOR, full 16-row truth table, demands at least one hidden neuron, solves quickly within the 5-minute backstop, clearly differs from `xor_classification` (4 rows) and `mnist_classification` (real-world 14×14 images). #262 lists 6-bit multiplexer and N-bit symmetry as alternatives; the implementer picks one and justifies the choice in the PR.
- **Re-use existing telemetry hooks.** The current code already uses `onTrainingEvent` to capture `bestFitness`/`averageFitness`/topology counts per generation; #263 simply adds an `accuracy` field. If `evolveDir` cannot expose enough data to record accuracy cleanly, the implementer falls back to chunk-boundary sampling and raises an upstream issue against `stSoftwareAU/NEAT-AI` (per the parent issue's instruction).
- **AGENTS.md update lands last.** Moving `adaptive_mutation` out of the exempt list before the rewrite is complete would be a false claim — review enforces the no-warm-start policy, so the metadata change must follow the code change.
- **No event-driven version.** The parent issue's note explicitly directs new examples toward batch (`evolveDir`) until upstream event-driven boilerplate improves.
- **Issue closure handled by the worker.** Per planning rules, this issue will be closed automatically; reopen it to iterate on the plan.

### Assumptions

- The chosen classification task can be solved by NEAT-AI from a uniform-random `new Creature(input, output)` seed within `timeoutMinutes: 5`. This is the parent issue's stated expectation; if it proves false the implementer raises a separate upstream NEAT-AI issue rather than weakening the demo (parent issue: ‘If the AI cannot find a good solution to a well-known problem, that is a genuine limitation worth raising as a separate issue’).
- No new shared `common/` utilities are needed — the classification task primitives are example-specific. If reuse value is identified during implementation, the implementer can promote them to `common/` in a follow-up.

This planning issue will be closed by the worker once the planning label is removed. **Reopen it** to request changes to the plan.

**Comment by stservice** [comment]
## Planning Complete

Planning complete. **3 sub-issue(s)** created:

- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/264
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/263
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/262

---
🤖 Processed by: stservice

**Comment by stservice** [comment]
Planning complete — sub-issues created.
